### PR TITLE
Some long reads tests using PacBio data.

### PR DESCRIPTION
**Pull request by cmnbroad**
The tests are written using the new plugin API (during which discovered that we need to add a utility method that resolves an index file name for a file that doesn't yet exist, given the file's name, but that can go into a separate PR). I have some ONT data as well but its currently too large to put in the repo.

### CODE REVIEW

File: src/test/java/htsjdk/samtools/LongReadsTest.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    @Test
+    final public void testQueryOverlappingForEachRead() throws IOException {
+        final IOPath copiedBAMPath = createBAMCopyWithIndex(PAC_BIO_READS);
+
+        // for each mapped read in the original source BAM, query both the original and the copy for
+        // all reads that overlap that read's coordinates, and compare the results
+        try (final ReadsDecoder sourceDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(ReadsBundle.resolveIndex(PAC_BIO_READS));
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/LongReadsTest.java:69]: This is an aside, but ReadsBundle.resolveIndex() probably needs a bit of renaming.  Seeing it in practice makes me think it should be ReadsBundle.withResolvedIndex() or something along those lines to make it clear that it's retuning a new reads bundle.

File: src/test/java/htsjdk/samtools/LongReadsTest.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+             final ReadsDecoder originalDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(ReadsBundle.resolveIndex(PAC_BIO_READS));
+             final ReadsDecoder copyDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(ReadsBundle.resolveIndex(tmpBAMPath))) {
+            for (final SAMRecord samRecord : sourceDecoder) {
+                //skip unmapped records since we're going to use the read's coordinates as a query
+                if (!samRecord.getReadUnmappedFlag()) {
+                    try (final CloseableIterator<SAMRecord> originalIt = originalDecoder.queryContained(
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/LongReadsTest.java:103]: We should probably add an overload of queryContained that takes a Locatable so you can just pass in the SamRecord.  Is it possible to make Locatable a subinterface of HtsInterval which returns smaller primitives?

File: src/test/java/htsjdk/samtools/LongReadsTest.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                new ReadsEncoderOptions().setPreSorted(false))) {
+            bamEncoder.setHeader(samFileHeader);
+            for (final SAMRecord samRecord : recordsInReverseOrder) {
+                bamEncoder.write(samRecord);
+            }
+        }
+
+        // now read back in both the original and the newly sorted reads that we wrote out, and compare the order
+        try (final ReadsDecoder originalDecoder =
+                     HtsDefaultRegistry.getReadsResolver().getReadsDecoder(
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/LongReadsTest.java:173]: I wonder if we should add methods to reduce the boiler plate.  I wish we had default values in java...

File: src/test/java/htsjdk/samtools/LongReadsTest.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                Assert.assertEquals(sortedIt.next().getAlignmentStart(), originalIt.next().getAlignmentStart());
+            }
+            Assert.assertFalse(sortedIt.hasNext());
+        }
+
+    }
+
+    private IOPath createBAMCopyWithIndex(final IOPath originalBAM) {
+        final IOPath tmpBAMPath = IOPathUtils.createTempPath("longReadsRoundTrip", ".bam");
+        // we need a better way to resolve the sibling - we can't use SamFiles.findIndex() because the file has
```
Comment (lbergelson) [inline src/test/java/htsjdk/samtools/LongReadsTest.java:196]: That's a good point...

**Review by lbergelson — APPROVED** [review]
👍 Looks good to me.  I have some comments unrelated to the tests.

File: src/test/java/htsjdk/samtools/LongReadsTest.java (hunk 1 above)
Comment (cmnbroad) [reply src/test/java/htsjdk/samtools/LongReadsTest.java:69]: Yes, makes sense.

File: src/test/java/htsjdk/samtools/LongReadsTest.java (hunk 2 above)
Comment (cmnbroad) [reply src/test/java/htsjdk/samtools/LongReadsTest.java:103]: The Locatable overload might be a good idea. Though we'd probably want to add parallel ones for query and queryContained. I think you'd get incompatible return types if you tried to make Locatable to extend HtsInterval though.

File: src/test/java/htsjdk/samtools/LongReadsTest.java (hunk 3 above)
Comment (cmnbroad) [reply src/test/java/htsjdk/samtools/LongReadsTest.java:173]: Yeah we should come up with something to make this easier.

File: src/test/java/htsjdk/samtools/LongReadsTest.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+            for (final SAMRecord samRecord : bamDecoder) {
+                bamEncoder.write(samRecord);
+            }
+        }
+        // mark the index for deletion too once its been created
+        tempIndexPath.toPath().toFile().deleteOnExit();
+        return tmpBAMPath;
+    }
+
+    private void compareQueryResults(final CloseableIterator<SAMRecord> originalIt, final CloseableIterator<SAMRecord> copyIt) {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/test/java/htsjdk/samtools/LongReadsTest.java:217]: Oh, yeah....

**Pull request merged**